### DISCUSSION TITLE
new perspective to CRNN

### DISCUSSION THREAD

**Discussion by yewalenikhil65** [DENG-MIT/CRNN#22] [Ideas]
@jiweiqi 
Here is something going on in my mind as i am trying to fit stiff systems.
we know that,
`1) 
X' = CRNN(X(t) ,t) , where CRNN= w_out*(exp.(w_in'* x + w_b))`
where ` w_out, w_in and w_b` are parameters.. Naturally, for large systems to be approximated, it is more difficult to train as there are so many parameters in these arrays.

`2) X' = w_out * oderatelaws = w_out*  [f(X,t) .*k] `  , where `w_out , k ` are parameters to be learnt (along with function `f`)
this is another way (although exactly same) to express ODEs of reactions.. Doesn't this mean there are less number of params to be dealt with as compared to first approach ? Is there any benefit by arranging system this way? If yes, is there any way to create a NN out of this ?

**Comment by jiweiqi** [comment]
Hi @yewalenikhil65 , this is a very brilliant idea. I think this discussion can also be formulated as "is it beneficial to relax part of the physics laws to trade better performance (accuracy, learnability, etc.)". I believe this is worthy and sometimes necessary.

This can be formulated as `X' = w_out * [NN(X) .*k]`, in which we don't use a polynomial form to represent the rate law. I actually did the same way in the yeast oscillation problem in this repo. If you are familiar with Quasi-Steady-State Assumptions, you will find that the reaction rate is usually in rational form, instead of a polynomial form. This is implicitly applied to most biology models. So, to your last question, this is fairly straightforward to implement. By the way, one can augment modern fancy NN architecture of autoencoder, graph neural network into CRNN in such away. We did try to initiate the discussions on how much physics we need in CRNN in the CRNN paper. The flexibility is indeed one of the most important motivations to propose the CRNN framework

The downside is that this will make the learned model depart a little bit from legacy forms. The good news is that the model still preserves quite a lot of interpretability from the information of `w_out`.

Above are my two cents. We should definitely continue this discussion. If you are excited about it, you can try it in the systems listed in this repo and the biomedical systems you are working on. I will be very happy to collaborate!

**Reply by jiweiqi** [reply]
* Training neural networks could be counterintuitive ahaha. I would expect that learning a regular NN is easier than learning a CRNN since NN has more parameters.
* For sure, one can choose any activation that fits, like tanh, gelu, relu, softplus
* the positivity is easy to be forced

**Reply by jiweiqi** [reply]
Here is the demo code from the yeast example
https://github.com/DENG-MIT/CRNN/blob/4fe08749f642677953c20d6d33c0c30079af9aa8/yeast-glycolysis/yeast_glycolysis.jl#L128-L158

**Reply by yewalenikhil65** [reply]
Directing you to https://github.com/yewalenikhil65/Application_of_CRNN/blob/main/README.md 
at the bottom,of read me file.. i have updated my say on this.. we can try at [here](https://github.com/yewalenikhil65/Application_of_CRNN) for now, and if successful add to your main CRNN repo.

**Reply by yewalenikhil65** [reply]
> Here is the demo code from the yeast example
> 
> https://github.com/DENG-MIT/CRNN/blob/4fe08749f642677953c20d6d33c0c30079af9aa8/yeast-glycolysis/yeast_glycolysis.jl#L129-L142

But `CRNN` approach is still used ? and `w_in_x` also ?

**Reply by yewalenikhil65** [reply]
Also, for `NN(X(t),t) = [((x1(t))^2) ; x1(t) ;  x3(t); x2(t)*x4(t)]` , these are set of polynomial and/or (rational functions ) in a vector. These are different **set of functions** .. so would 1 NN be enough to approximate a such set of functions ?

**Reply by yewalenikhil65** [reply]
@jiweiqi another observation.. where is `rep = re(pnn) ` being used in later `predict_neuralode`. It seems like inner NN used to approximate oderatelaws.. but i am unable to see where they are being used further
```julia
function predict_neuralode(u0, p, batch=ntotal) 
     pcrnn = p[1:np] 
     pnn = p[np + 1:end] 
     global w_in, w_b, w_out, w_J = p2vec(pcrnn) 
     global rep = re(pnn) 
     prob_ = remake(prob, u0=u0, p=pcrnn) 
     pred = clamp.(Array(solve(prob_, alg, saveat=tsteps[1:batch], sensalg=sense)), lb, ub) 
     return pred 
 end
```

**Reply by jiweiqi** [reply]
It is inside the crnn function.

**Reply by yewalenikhil65** [reply]
@jiweiqi , is `batchsize` that is later used as `sample` kind of training over only partial data in `yeast` and `robers` problem ? (ass diffeqflux documentation suggests in many instances to train in partial time-spans to avoid local minima)

**Reply by jiweiqi** [reply]
Yes, you can think about it as training partial of data. There is a [paper](https://arxiv.org/pdf/2006.10711.pdf) talking about such kind of approach although it is not always useful in practice.

**Reply by yewalenikhil65** [reply]
> Yes, you can think about it as training partial of data. There is a [paper](https://arxiv.org/pdf/2006.10711.pdf) talking about such kind of approach although it is not always useful in practice.

there is ongoing PR at DiffEqFlux.. called multiple shooting. Someone suggested me to do this for increasing performance for stiff problems. but couldn't implement. This one seems to be in order.
https://github.com/Manas2030/DiffEqFlux.jl/blob/multiple_shooting/src/multiple_shooting.jl

This too is a partial training of data.. But your approach seemed more of `changing` the span of training data.. while this one divides the span into chunks.. To what effect, i am not sure.

**Reply by jiweiqi** [reply]
Looks interesting.

**Reply by yewalenikhil65** [reply]
Would you be interested in a google meet someday to discuss more in detail. It's bit straining writing everything

**Reply by jiweiqi** [reply]
Sure, that's a good idea. Let's do it sometime next week

**Reply by yewalenikhil65** [reply]
> Sure, that's a good idea. Let's do it sometime next week

Let me know time for online meet that will be convenient for you..on email..It would be nice to add my supervisor also formally

**Reply by jiweiqi** [reply]
https://www.when2meet.com/?11329788-IDATF Let's use this one to schedule a time
